datashader/data_libraries/dask.py

```diff
@@ -70,18 +76,103 @@ def default(glyph, df, schema, canvas, summary, cuda=False):
     y_mapper = canvas.y_axis.mapper
     extend = glyph._build_extend(x_mapper, y_mapper, info, append)
 
-    def chunk(df):
+    # Here be dragons
```

Comment:
Agreed, but explaining precisely what those dragons are would be helpful! :-)

Reply:
Low-level dask graphs can represent a very broad range of operations and data.  Dask Arrays and Dataframes are simply metadata describing the layer of an associated Low-Level Dask Array. For example:

```python
>>> dsk = {
      ('a', 0, 0): (np.full, (2, 2), 1),
      ('a', 0, 1): (np.full, (2, 2), 2),
      ('a', 1, 0): (np.full, (2, 2), 3),
      ('a', 1, 1): (np.full, (2, 2), 4)}

>>> chunks = ((2, 2), (2, 2))
>>> array = da.Array(dsk, 'a', chunks, meta=np.empty((0,0), dtype=np.int32))
>>> array
dask.array<a, shape=(4, 4), dtype=int32, chunksize=(2, 2), chunktype=numpy.ndarray>
>>> array.compute()                                                                                                                                                      
Out[7]: 
array([[1, 1, 2, 2],
       [1, 1, 2, 2],
       [3, 3, 4, 4],
       [3, 3, 4, 4]])
```

In the above example, the low-level graph, **chunks** describe the purported size of the arrays produced by the 4 dask tasks, while the **meta** describes the **type** of each chunk, here represented with an empty array. This is to support arrays other that NumPy (e.g. [sparse](https://sparse.pydata.org) arrays). At low-level graph

The dragon is that it is entirely possibly to construct **dsk** in a way that disagrees with the metadata and *this doesn't matter at all at the graph level*. The only time there's a problem is:

1. if ``array.compute()`` is directly called (Because dask will try to construct an array of type **meta**)
2. If the output of ``arrays`` tasks are used as input to tasks that expect ndarrays.

(1) doesn't apply in this case  and here, a dask Array is constructed from tasks producing Pandas Dataframes which are passed into tasks (combine, aggregate),  that expect Dataframes, circumventing (2).

The final dask Array produced by the reduction describes a task that does return an ndarray so everything works out, under the hood.

Reply:
Right, so in summary, no issues in the way it's currently used in the codebase (given that everything goes via dataframes), just leaving a signpost for future developers to look out for if they start hacking the code for another purpose?

Reply:
Yep, I can leave a link to the explanation in the code.